crypto/ripemd160.c

```diff
@@ -1,343 +1,344 @@
-    low  = ( ctx->total[0] <<  3 );
+    /* Final mixing stage */
+    T = self->h[1] + CL + DR;
+    self->h[1] = self->h[2] + DL + ER;
+    self->h[2] = self->h[3] + EL + AR;
+    self->h[3] = self->h[4] + AL + BR;
+    self->h[4] = self->h[0] + BL + CR;
+    self->h[0] = T;
+
+    /* Clear the buffer and wipe the temporary variables */
+    T = AL = BL = CL = DL = EL = AR = BR = CR = DR = ER = 0;
+    memset(&self->buf, 0, sizeof(self->buf));
+    self->bufpos = 0;
+}
 
-    PUT_UINT32_LE( low,  msglen, 0 );
-    PUT_UINT32_LE( high, msglen, 4 );
+void ripemd160_process(ripemd160_state * self, const uint8_t *p, size_t length)
+{
+    unsigned long bytes_needed;
+
+    /* Some assertions */
+    assert(p != NULL);
+
+    /* We never leave a full buffer */
+    assert(self->bufpos < 64);
+
+    while (length > 0) {
+        /* Figure out how many bytes we need to fill the internal buffer. */
+        bytes_needed = 64 - self->bufpos;
+
+        if ((unsigned long) length >= bytes_needed) {
+            /* We have enough bytes, so copy them into the internal buffer and run
+             * the compression function. */
+            memcpy(&self->buf.b[self->bufpos], p, bytes_needed);
+            self->bufpos += bytes_needed;
+            self->length += bytes_needed << 3;    /* length is in bits */
+            p += bytes_needed;
+            ripemd160_compress(self);
+            length -= bytes_needed;
+            continue;
+        }
+
+        /* We do not have enough bytes to fill the internal buffer.
+         * Copy what's there and return. */
+        memcpy(&self->buf.b[self->bufpos], p, length);
+        self->bufpos += length;
+        self->length += length << 3;    /* length is in bits */
+        return;
+    }
+}
 
-    last = ctx->total[0] & 0x3F;
-    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );
+void ripemd160_done(ripemd160_state * self, uint8_t out[RIPEMD160_DIGEST_LENGTH])
+{
+    /* Append the padding */
+    self->buf.b[self->bufpos++] = 0x80;
 
-    ripemd160_Update( ctx, ripemd160_padding, padn );
-    ripemd160_Update( ctx, msglen, 8 );
+    if (self->bufpos > 56) {
+        self->bufpos = 64;
+        ripemd160_compress(self);
+    }
 
-    PUT_UINT32_LE( ctx->state[0], output,  0 );
-    PUT_UINT32_LE( ctx->state[1], output,  4 );
-    PUT_UINT32_LE( ctx->state[2], output,  8 );
-    PUT_UINT32_LE( ctx->state[3], output, 12 );
-    PUT_UINT32_LE( ctx->state[4], output, 16 );
+    /* Append the length */
+    self->buf.w[14] = (uint32_t) (self->length & 0xFFFFffffu);
+    self->buf.w[15] = (uint32_t) ((self->length >> 32) & 0xFFFFffffu);
+#ifdef PCT_BIG_ENDIAN
+    byteswap32(&self->buf.w[14]);
+    byteswap32(&self->buf.w[15]);
+#endif
+    self->bufpos = 64;
+    ripemd160_compress(self);
 
-    memzero(ctx, sizeof(RIPEMD160_CTX));
+    /* Copy the final state into the output buffer */
+#ifdef PCT_BIG_ENDIAN
+    byteswap_digest(self->h);
+#endif
+    memcpy(out, &self->h, RIPEMD160_DIGEST_SIZE);
```

Comment:
Let's `memzero()` the context when done, like we do in the SHA implementations. It looks like we are not doing it in blake though.

Reply:
Done in https://github.com/trezor/trezor-firmware/pull/4107/commits/704bc4f1dc9e804e3e9c095e3dab1d610a7636e6.